libraries/standard/mqtt/integration-test/mqtt_system_test.c

```diff
@@ -848,3 +877,68 @@ void test_MQTT_ProcessLoop_KeepAlive( void )
     elapsedTime = context.lastPacketTime - connectPacketTime;
     TEST_ASSERT_LESS_OR_EQUAL( MQTT_KEEP_ALIVE_INTERVAL_SECONDS * 1500, elapsedTime );
 }
+
+/**
+ * @brief Verifies the behavior of the MQTT library in a restored session connection with the broker.
+ * Tests that the library resends PUBREL packets to the broker in a restored session for an incomplete
+ * PUBLISH operation in a previous connection.
+ */
+void test_MQTT_Connect_Restore_Session( void )
+{
+    /* Terminate TLS session and TCP connection network connection to discard current MQTT session
+     * that was created as a "clean session". */
+    ( void ) Openssl_Disconnect( &networkContext );
+
+    /* Establish a new MQTT connection over TLS with the broker with the "clean session" flag set to 0
+     * to start a persistent session with the broker. */
+
+    /* Create the TLS+TCP connection with the broker. */
+    TEST_ASSERT_EQUAL( OPENSSL_SUCCESS, Openssl_Connect( &networkContext,
+                                                         &serverInfo,
+                                                         &opensslCredentials,
+                                                         TRANSPORT_SEND_RECV_TIMEOUT_MS,
+                                                         TRANSPORT_SEND_RECV_TIMEOUT_MS ) );
+    TEST_ASSERT_NOT_EQUAL( -1, networkContext.socketDescriptor );
+    TEST_ASSERT_NOT_NULL( networkContext.pSsl );
+
+    /* Establish a new MQTT connection for a persistent session with the broker. */
+    establishMqttSession( &context, &networkContext, false, &persistentSession );
+    TEST_ASSERT_FALSE( persistentSession );
+
+    /* Publish to a topic with Qos 2. */
+    TEST_ASSERT_EQUAL( MQTTSuccess, publishToTopic(
+                           &context, TEST_MQTT_TOPIC, MQTTQoS2 ) );
+
+    /* Disconnect on receiving PUBREC so that we are not able to complete the QoS 2 PUBLISH in the current connection. */
+    disconnectOnPacketType = MQTT_PACKET_TYPE_PUBREC;
+    TEST_ASSERT_EQUAL( MQTTSendFailed,
+                       MQTT_ProcessLoop( &context, 2 * MQTT_PROCESS_LOOP_TIMEOUT_MS ) );
+    TEST_ASSERT_FALSE( receivedPubComp );
```

Comment:
This test, covers the path for re sending the PUBREL ack, there are a few more cases to be covered for restoring the session.
1. Resend Unacked PUBLISHES with dup flag (QoS1 and QoS2)
2. Receive duplicate PUBREL (QoS2)
3. Receive duplicate PUBLISHES(QoS1 and QoS2)

Reply:
1. How would we test that the DUP flag is set? I don't think we have access to the publish packet that's created.
2. Isn't that testing more that the broker is up to spec rather than our client? Though I agree we should test it for the purpose of ensuring the state engine works fine.

Reply:
1. I believe you are asking about DUP flag for re sending the unacked publishes. It will be application's responsibility to do that. In this case, integration test will be able to do that.
2. Yea. We have logic for handling the states when duplicate publishes or acks are received. We would need to test that in our integration tests.

Reply:
Will take care of the other cases of restore session logic in succeeding PRs

Reply:
Added a test for case 2. of receiving duplicate `PUBREL` packet in restored session